notebooks/radp_library.py

```diff
@@ -867,3 +941,242 @@ def rfco_to_best_server_shapes(
+
+
+def plot_ue_tracks_side_by_side(df1: pd.DataFrame, df2: pd.DataFrame) -> None:
+    """
+    Plots the movement tracks of unique UE IDs from two DataFrames side by side.
+    """
+    # Set up subplots with 2 columns for side by side plots
+    fig, axes = plt.subplots(1, 2, figsize=(25, 10))  # 2 rows, 2 columns (side by side)
+
+    # Plot the first DataFrame
+    plot_ue_tracks_on_axis(df1, axes[0], title="DataFrame 1")
+
+    # Plot the second DataFrame
+    plot_ue_tracks_on_axis(df2, axes[1], title="DataFrame 2")
+
+    # Adjust layout and show
+    plt.tight_layout()
+    plt.show()
+
+
+def plot_ue_tracks_on_axis(df: pd.DataFrame, ax, title: str) -> None:
+    """
+    Helper function to plot UE tracks on a given axis.
+    """
+    data = df
+    unique_ids = data["mock_ue_id"].unique()
+    num_plots = len(unique_ids)
+
+    color_map = cm.get_cmap("tab20", num_plots)
+
+    for idx, ue_id in enumerate(unique_ids):
+        ue_data = data[data["mock_ue_id"] == ue_id]
+
+        for i in range(len(ue_data) - 1):
+            x_start = ue_data.iloc[i]["lon"]
+            y_start = ue_data.iloc[i]["lat"]
+            x_end = ue_data.iloc[i + 1]["lon"]
+            y_end = ue_data.iloc[i + 1]["lat"]
+
+            dx = x_end - x_start
+            dy = y_end - y_start
+            ax.quiver(
+                x_start,
+                y_start,
+                dx,
+                dy,
+                angles="xy",
+                scale_units="xy",
+                scale=1,
+                color=color_map(idx),
+            )
+
+        ax.scatter(
+            ue_data["lon"], ue_data["lat"], color=color_map(idx), label=f"UE {ue_id}"
+        )
+
+    ax.set_title(title)
+    ax.legend()
+
+
+def calculate_distances_and_velocities(group: pd.DataFrame) -> pd.DataFrame:
+    """Calculating distances and velocities for each UE based on sorted data by ticks."""
+    group["prev_longitude"] = group["lon"].shift(1)
+    group["prev_latitude"] = group["lat"].shift(1)
+    group["distance"] = group.apply(
+        lambda row: GISTools.get_log_distance(
+            row["prev_latitude"], row["prev_longitude"], row["lat"], row["lon"]
+        )
+        if not pd.isna(row["prev_longitude"])
+        else 0,
+        axis=1,
+    )
+    # Assuming time interval between ticks is 1 unit, adjust below if different
+    group["velocity"] = (
+        group["distance"] / 1
+    )  # Convert to m/s by dividing by the seconds per tick, here assumed to be 1s
+    return group
+
+
+def preprocess_ue_data(df: pd.DataFrame) -> pd.DataFrame:
+    """Preprocess the UE data by calculating distances and velocities."""
+    df.sort_values(by=["mock_ue_id", "tick"], inplace=True)
+
+    df = df.groupby("mock_ue_id").apply(calculate_distances_and_velocities)
+    # Drop the temporary columns
+    df.drop(["prev_longitude", "prev_latitude"], axis=1, inplace=True)
+
+    return df
+
+
+def get_predicted_alpha(data: pd.DataFrame, alpha0: float,seed: int) -> float:
```

Comment:
"""This is the main API entry point for alpha prediction, given UE tracks.

@param data : provide columns
"""